pkg/lightning/backend/remote/chunk_sender.go

```diff
@@ -0,0 +1,396 @@
+
+		nextChunkID := result.HandledChunkID + 1
+		for nextChunkID <= expectedChunkID {
+			url := fmt.Sprintf(putChunkURL, c.e.addr, c.e.clusterID, c.e.loadDataTaskID, c.id, nextChunkID)
+
+			buf, err := c.chunksCache.get(nextChunkID)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			c.e.logger.Info("retry to put chunk",
+				zap.Uint64("sender", c.id),
+				zap.Uint64("chunkID", nextChunkID))
+
+			data, err := sendRequest(ctx, c.httpClient, "PUT", url, buf)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			result = new(PutChunkResult)
+			err = json.Unmarshal(data, result)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			nextChunkID = result.HandledChunkID + 1
+		}
+	}
+
+	state := c.state.Load().(*chunksState)
+	lastFlushedChunkID := state.FlushedChunkID + 1
+	for lastFlushedChunkID <= result.FlushedChunkID {
+		err := c.chunksCache.clean(lastFlushedChunkID)
+		if err != nil {
+			c.e.logger.Info("failed to clean chunk cache",
+				zap.Uint64("chunkID", lastFlushedChunkID),
+				zap.Error(err))
+		}
+		lastFlushedChunkID++
+	}
+
+	c.state.Store(&chunksState{
+		FlushedChunkID: result.FlushedChunkID,
+		HandledChunkID: result.HandledChunkID,
+	})
+	return nil
+}
+
+func (c *chunkSender) sendFlushToRemote(ctx context.Context) error {
+	for {
+		url := fmt.Sprintf(flushURL, c.e.addr, c.e.clusterID, c.e.loadDataTaskID, c.id)
+
+		data, err := sendRequest(ctx, c.httpClient, "POST", url, nil)
+		if err != nil {
+			c.e.logger.Error("failed to flush", zap.Error(err))
+			return errors.Trace(err)
+		}
+
+		result := new(FlushResult)
+		err = json.Unmarshal(data, result)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		if result.Canceled || result.Error != "" {
+			c.e.logger.Error("failed to flush",
+				zap.Bool("canceled", result.Canceled),
+				zap.String("error", result.Error))
+			return common.ErrLoadDataTaskCanceled.FastGenByArgs(c.e.loadDataTaskID, result.Error)
+		}
+
+		if result.Finished {
+			c.e.logger.Info("load data task finished")
+			return nil
+		}
+
+		// make sure all chunks are flushed
+		flushedChunkID := result.FlushedChunkIDs[c.id]
+		if flushedChunkID == c.getLastChunkID() {
+			// all chunks are flushed, clean cache and update state
+			state := c.state.Load().(*chunksState)
+			lastFlushedChunkID := state.FlushedChunkID + 1
+			for lastFlushedChunkID <= flushedChunkID {
+				err := c.chunksCache.clean(lastFlushedChunkID)
+				if err != nil {
+					c.e.logger.Warn("failed to clean chunk cache", zap.Uint64("chunkID", lastFlushedChunkID), zap.Error(err))
+				}
+				lastFlushedChunkID++
+			}
+			state.FlushedChunkID = flushedChunkID
+
+			c.state.Store(state)
+			return nil
+		}
+
+		// the flushed chunkID is not equal to the last sent chunkID,
+		// because the remote worker may restart, retry to put chunk.
+		err = c.putEmptyChunk(ctx)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+}
+
+func (c *chunkSender) close() error {
+	// notify chunkSenderLoop to quit
```

Comment:
better to derive a context using WithCancel to quit. Because not all functions in `chunkSenderLoop` will check `loopQuitChan`, so the cancel may not be timely.